### PR TITLE
Complete support for ARG statement handling

### PR DESCRIPTION
**Pull request by iciclespider**
This started with the branch @nishakm submitted, but adds complete support for ARGs usage in all the same places ENVs are currently supported. It also enhanced all ENV pytests to also test ARGs, and ensured pytests have 100% coverage.

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered

### CODE REVIEW

File: dockerfile_parse/parser.py [hunk 1]
```diff
@@ -360,7 +395,14 @@ def baseimage(self, new_image):
...
         """
-        images = self.parent_images or [None]
+        images = []
+        for instr in self.structure:
+            if instr['instruction'] == 'FROM':
+                image, _ = image_from(instr['value'])
+                if image is not None:
+                    images.append(image)
+        if not images:
+            raise RuntimeError('No stage defined to set base image on')
```
Comment (MartinBasti) [inline dockerfile_parse/parser.py:405]: Why this  change is needed?

File: dockerfile_parse/parser.py [hunk 2]
```diff
@@ -15,72 +15,87 @@
...
+    """
+    parser_attr = 'args'
 
 
 class DockerfileParser(object):
     def __init__(self, path=None,
                  cache_content=False,
                  env_replace=True,
                  parent_env=None,
+                 build_args=None,
```
Comment (MartinBasti) [inline dockerfile_parse/parser.py:90]: This has to be appended as last param, otherwise we may break BW compatibility for user who are just using it as positional args (I'd wish having `*` in params list in py2)

File: dockerfile_parse/util.py [hunk 3]
```diff
@@ -46,13 +46,16 @@ class WordSplitter(object):
     SQUOTE = "'"
     DQUOTE = '"'
 
-    def __init__(self, s, envs=None):
+    def __init__(self, s, args=None, envs=None):
```
Comment (MartinBasti) [inline dockerfile_parse/util.py:49]: Waiving BW compat, this is just util module

File: test.sh [hunk 4]
```diff
@@ -71,7 +69,7 @@ function setup_dfp() {
 
   # CentOS needs to have setuptools updates to make pytest-cov work
   if [[ $OS != "fedora" ]]; then
-    $RUN $PIP install -U setuptools
+    $RUN $PIP install -U 'setuptools<45'
```
Comment (ben-alkov) [inline test.sh:72]: You don't have to pin to a specific version here. As long as you

```bash
  $RUN $PIP install -U pip
  $RUN $PIP install -U setuptools
```

pip and setuptools will do the right thing

**Comment by iciclespider** [comment]
I added a `test-test.sh` file which exercises all uses of `test.sh` to verify that my `test.sh` changes complete.

File: test.sh (hunk 4 above)
Comment (MartinBasti) [reply test.sh]: because old pip cannot understand supported python version listed in packages. Upgraded pip will install setuptools=44 under py2 as expected

File: test.sh [hunk 5]
```diff
@@ -94,7 +99,7 @@ case ${ACTION} in
   ;;
 "pylint")
   setup_dfp
-  $RUN $PKG install -y "${PYTHON}-pylint"
+  $RUN $PIP install pylint
```
Comment (MartinBasti) [inline test.sh:102]: We used distro specific version intentionally. Upstream pylint changes too often and causes many false positives we have to handle.

File: test-test.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
```
Comment (MartinBasti) [inline test-test.sh:1]: Please drop this script from PR.

If needed travis CI should be updated, there are specific requirements, we don't test everything.

* latest bandit once per py2/py3 (it's not needed for every distro)
* markdownlint is enough to run just once too
* pylint py2, freezed at F29 as newer pylint version stopped supporting py2
* pylint py3, tested on latest fedora only, no need to test each distro
* F32 should be able run only py3, we will replace F30 with F32 soon

File: test.sh (hunk 4 above)
Comment (ben-alkov) [reply test.sh]: @iciclespider I just had to handle this exact issue in [osbs-client](https://github.com/containerbuildsystem/osbs-client/pull/984/files) - as long as you `$RUN $PIP install -U pip`, pip will do the right thing and install[1] a correct, working setuptools for the OS/Python it's running under.

[1] I did `$RUN $PIP install -U setuptools` immediately after updating pip; YMMV

File: test.sh (hunk 5 above)
Comment (iciclespider) [reply test.sh:102]: > Upstream pylint changes too often and causes many false positives we have to handle.

Would not this be better addressed by pip installing a specific version, rather then relying on the fact that the distro packaging is somewhat behind the latest?

File: test-test.sh (hunk 6 above)
Comment (iciclespider) [reply test-test.sh]: Removed.

File: test.sh (hunk 4 above)
Comment (iciclespider) [reply test.sh]: Oh, I see what you are saying. Added the pip upgrade before the setuptools upgrade.

File: test.sh (hunk 5 above)
Comment (MartinBasti) [reply test.sh:99]: It happens that when installing from pip some dependencies were broken, or incompatible dependencies were installed. We would need to freeze pip versions. So for me it's easier to rely on distro installations that resolves dependency issues for me.

**Comment by MartinBasti** [comment]
Can you please squash last 2 commits into `Complete support for ARG statement handling` commit please?

File: python-dockerfile-parse.spec [hunk 7]
```diff
@@ -121,7 +121,7 @@ py.test-%{python3_version} -v tests
 * Tue Jun 02 2020 Robert Cerven <rcerven@redhat.com> 0.0.18-1
 - new upstream release: 0.0.18
 
-* Fri Apr 24 2020 Martin Bašti <mbasti@redhat.com> 0.0.17-1
+* Fri Apr 24 2020 Martin Basti <mbasti@redhat.com> 0.0.17-1
```
Comment (MartinBasti) [inline python-dockerfile-parse.spec:124]: Pls add this change to separate commit

File: test.sh [hunk 8]
```diff
@@ -23,6 +24,16 @@ for dir in ${EXTRA_MOUNT:-}; do
   engine_mounts=("${engine_mounts[@]}" -v "$dir":"$dir":z)
 done
 
+# Force recreation of the container
```
Comment (MartinBasti) [inline test.sh:27]: test.sh changes are also unrelated to feature, can you extract it into separate commit?

Comment (iciclespider) [reply test.sh:27]: By "separate commit", do you mean a different Pull Request? Or a separate commit within this Pull Request?

File: python-dockerfile-parse.spec (hunk 7 above)
Comment (iciclespider) [reply python-dockerfile-parse.spec:124]: By "separate commit", do you mean a different Pull Request? Or a separate commit within this Pull Request?

**Comment by iciclespider** [comment]
I reverted the `test.sh` changes to be the bare minimum needed to get this PR to pass the Travis CI build.

**Comment by MartinBasti** [comment]
@athos-ribeiro @chmeliik @rcerven @ben-alkov PTAL

**Comment by MartinBasti** [comment]
Release notes:
- Added support for ARG statement handling
